### PR TITLE
Feature: certificate implementation

### LINKED ISSUE #1165: Batch Account Certificates Experience
**Description**
No certificate functionality exists. It would be good to add at least add/view/delete operations for certs for an account.

### PR DESCRIPTION
**Pull request by johnnyzhang82**
Fix #1165. npmImplement certificate interface for BatchLab including:

1. List/Get certificate
2. Create certificate
3. Delete certificate
4. Cancel deletion of a certificate

image

### CODE REVIEW

File: app/app.routes.ts [hunk 1]
```diff
@@ -73,6 +75,15 @@ export const routes: Routes = [
             { path: ":id", component: PoolDetailsComponent }, // pools/{pool.id}
         ],
     },
+    {
+        path: "certificates",
+        canActivate: [NavigationGuard],
+        component: CertificateHomeComponent,
+        children: [
+            { path: "", component: CertificateDefaultComponent }, // thumbprint/
```
Comment (timotheeguerin) [inline app/app.routes.ts]: comment should be `certificates/` no?

File: package.json [hunk 2]
```diff
@@ -155,6 +155,7 @@
     "mousetrap": "^1.6.0",
     "ms-rest-js": "~0.2.3",
     "node-fetch": "~1.7.3",
+    "node-forge": "^0.7.4",
```
Comment (timotheeguerin) [inline package.json]: I just registered for OSS approval

File: app/services/certificate.service.ts [hunk 3]
```diff
@@ -0,0 +1,210 @@
+import { Injectable } from "@angular/core";
+import { List } from "immutable";
+import { AsyncSubject, Observable, Subject } from "rxjs";
+// tslint:disable-next-line:no-var-requires
+const forge = require("node-forge");
```
Comment (timotheeguerin) [inline app/services/certificate.service.ts]: es6 imports

File: app/services/certificate.service.ts [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+        };
+        return subject.asObservable();
+    }
+
+    /**
+     * Helper function to get certificate extension string
+     * @param file
+     */
+    public getCertificateExtension(file: File) {
+        return file.name.substr(file.name.length - 3, 3).toLowerCase();
```
Comment (timotheeguerin) [inline app/services/certificate.service.ts]: Use `FileUrlUtils#getFileExtension()`

File: app/services/certificate.service.ts [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+            },
+            listNext: (client, nextLink: string) => client.certificate.listNext(nextLink),
+        });
+    }
+
+    public get basicProperties(): string {
+        return this._basicProperties;
+    }
+
+    public list(options?: any, forceNew?: boolean);
```
Comment (timotheeguerin) [inline app/services/certificate.service.ts]: is there actually a listNext for certificates?

File: app/components/certificate/action/reactivate/reactivate-certificate-dialog.html [hunk 6]
```diff
@@ -0,0 +1,3 @@
+<bl-simple-form [submit]="ok" [containerRef]="dialogRef" [multiUse]="false" size="small" actionName="Enable" title="Reactivate certificate">
+    <p>Do you want to reactivate the certificate'<b>{{certificateThumbprint}}</b>'?</p>
```
Comment (ascobie) [inline app/components/certificate/action/reactivate/reactivate-certificate-dialog.html]: need space or ': ' between here: 

certificate'<b>{{certificateThumbprint}}</b>'

File: app/components/certificate/browse/certificate-list.component.ts [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+import { List } from "immutable";
+import { Observable, Subscription } from "rxjs";
+
+import { Filter, FilterMatcher, Operator, autobind } from "@batch-flask/core";
+import { ListBaseComponent, ListSelection } from "@batch-flask/core/list";
+import { BackgroundTaskService } from "@batch-flask/ui/background-task";
+import { ContextMenu, ContextMenuItem } from "@batch-flask/ui/context-menu";
+import { LoadingStatus } from "@batch-flask/ui/loading";
+import { QuickListItemStatus } from "@batch-flask/ui/quick-list";
+import { Certificate, CertificateState } from "app/models";
```
Comment (ascobie) [inline app/components/certificate/browse/certificate-list.component.ts]: I have been adding a line break between @batch-flask imports and the other app imports. I think it makes it cleaner and easier to read.

File: app/components/certificate/browse/certificate-list.component.ts [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+                return QuickListItemStatus.normal;
+        }
+    }
+
+    public certificateStatusText(certificate: Certificate): string {
+        switch (certificate.state) {
+            case CertificateState.active:
+                return "";
+            default:
+                return `Certficate is ${certificate.state}`;
```
Comment (ascobie) [inline app/components/certificate/browse/certificate-list.component.ts]: Certificate

File: app/components/certificate/details/certificate-details.component.ts [hunk 9]
```diff
@@ -0,0 +1,108 @@
+import { ChangeDetectionStrategy, Component, OnDestroy, OnInit } from "@angular/core";
+import { MatDialog, MatDialogConfig } from "@angular/material";
+import { ActivatedRoute, Router } from "@angular/router";
+import { autobind } from "@batch-flask/core";
+import { ElectronRemote } from "@batch-flask/ui";
+import { Observable, Subscription } from "rxjs";
```
Comment (ascobie) [inline app/components/certificate/details/certificate-details.component.ts]: move flask imports out of this block

File: app/components/certificate/browse/certificate-list.component.ts (hunk 7 above)
Comment (timotheeguerin) [reply app/components/certificate/browse/certificate-list.component.ts]: @batch-flask imports can be added to the external imports above, as its techincally supposed to act like an external library

File: app/components/certificate/details/certificate-details.component.ts (hunk 9 above)
Comment (timotheeguerin) [reply app/components/certificate/details/certificate-details.component.ts]: i think they belong here

Comment (ascobie) [reply app/components/certificate/details/certificate-details.component.ts]: ok cool. if you are happy then i am also :)

File: app/services/certificate.service.ts (hunk 5 above)
Comment (johnnyzhang82) [reply app/services/certificate.service.ts]: Yes, there is nextLink in certificateListResult
image

**Pull request merged**